client/src/components/Draftail/decorators/Link.js

```diff
@@ -104,13 +121,84 @@ export const onPasteLink = (text, html, editorState, { setEditorState }) => {
       undefined,
       entityKey,
     );
-    nextState = EditorState.push(editorState, content, 'insert-characters');
-  } else {
-    nextState = RichUtils.toggleLink(editorState, selection, entityKey);
+    return EditorState.push(editorState, content, 'insert-characters');
+  }
+  return RichUtils.toggleLink(editorState, selection, entityKey);
+};
+
+/**
+ * Insert the pasted HTML or text, auto-linking URLs and emails.
+ */
+const insertContentWithLinks = (editorState, htmlOrText) => {
+  const selection = editorState.getSelection();
+  let content = editorState.getCurrentContent();
+
+  const { contentBlocks, entityMap } = convertFromHTML(htmlOrText);
+  const blockMap = ContentState.createFromBlockArray(
+    contentBlocks,
+    entityMap,
+  ).getBlockMap();
+
+  const blocks = blockMap.map((block) => {
+    const blockText = block.getText();
+    const pattern = new RegExp(linkPatternSource, 'ig');
+    // Find matches in the block, confirm the URL, create the entity, store the range.
+    const matches = Array.from(blockText.matchAll(pattern), (match) => {
+      // Account for punctuation chars valid in URLs but unlikely to be intended.
+      // For example "Go to https://example.com."
+      // Terminal Punctuation class: see https://www.unicode.org/review/pr-23.html.
+      const cleanURLPattern = match[1].replace(
+        /\p{Terminal_Punctuation}$/u,
```

Comment:
If we're doing this, I think we should consume all the trailing punctuations rather than just the last one, i.e. similar to Python's `rstrip()`. Otherwise, if we copy the following:

```
The roadmap here https://github.com/wagtail/roadmap...
but it's also here https://wagtail.org/roadmap?!?!
```

We'd get

> The roadmap here [https://github.com/wagtail/roadmap..](https://github.com/wagtail/roadmap..).
> but it's also here [https://wagtail.org/roadmap?!?](https://wagtail.org/roadmap?!?)!

Whereas if we consume all the trailing punctuations, i.e.

```suggestion
        /\p{Terminal_Punctuation}+$/u,
```

We'd get

> The roadmap here [https://github.com/wagtail/roadmap](https://github.com/wagtail/roadmap)...
> but it's also here [https://wagtail.org/roadmap](https://wagtail.org/roadmap)?!?!

Which I think is the common behaviour on various places. For example, on GitHub when pasting as-is you'd get:

The roadmap here https://github.com/wagtail/roadmap...
but it's also here https://wagtail.org/roadmap?!?!

Also on Slack:

<img width="451" alt="image" src="https://github.com/user-attachments/assets/c3085eb5-beb5-45dc-ae93-fb29510b6cb3">